### PR TITLE
[DX-658] Integrate API Security Best Practice Contributed By Dave Garvey

### PR DESCRIPTION
**Pull request by dcs3spp**
[DX-658](https://tyktech.atlassian.net/browse/DX-658)

[preview](https://deploy-preview-3216--tyk-docs.netlify.app/docs/nightly/apim-best-practice/overview)

David Garvey has kindly provided documentation for API Security Practice. The requirement is that this should be integrated into the best practices section of the new website. The link to the original thread is [here](https://tyktech.slack.com/archives/CKLM28JV6/p1692346259868339)

[DX-658]: https://tyktech.atlassian.net/browse/DX-658?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[The Confluence Article](https://tyktech.atlassian.net/wiki/spaces/RD/pages/2137718879/API+Security+Best+Practice)

### CODE REVIEW

File: tyk-docs/content/api-management/manage-apis/api-security/overview.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+title: "API Security Overview"
+date: 2020-09-01
+tags: ["API Security", "Authorisation", "Authentication", "Resource Consumption"]
+description: "API security best practices"
```
Comment (letzya) [inline tyk-docs/content/api-management/manage-apis/api-security/overview.md]: We need a slightly longer description for Google to attract our readers

**Review by letzya — COMMENTED** [review]
@dcs3spp what is the reason this article is split into multiple pages?

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+The gateway can also perform object property level authorisation, by validating that the values of the body data fields, `name` and `id`, conform to a schema.
+
+Lastly, the API handles object level authorisation by using custom logic. This typically involves using the value of the `authorization` header in combination with the ownership and authorisation model specific to the API to determine if the client is authorised to write the requested data.
+
+### GraphQL API - Querying Data
+
+
+{{< img src="/img/api-management/security/graphql-query-data.jpeg" alt="Rest API - Write Data" width="150px" >}}
+
+The client sends a _POST_ request using the path `/graphql` and body data containing a GraphQL query. The path `/graphql` is static and requires function level authorisation. The GraphQL query contains several elements:
```
Comment (davegarvey) [inline tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: The formatting for the HTTP method should be as code, as per previous entries.

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+- `name`: A standard object field, referring to a property of the profile object type. This requires object property authorisation.
+
+Next, the Gateway handles function level authorisation, by checking that the path, in the case `/graphql`, is authorised for access. It does this by cross referencing the security policies connected to the API key provided in the `authorization` header. Due to the nature of GraphQL using just a single endpoint, there is no need for additional path-based authorisation features, only a basic security policy is required.
+
+Another difference between this and the REST examples is in the way that the body data is authorised:
+
+All object types and fields contained in the query are checked against the API’s GraphQL schema, to ensure they are valid. In this case, the object type is `profile`, and the fields are `id` and `name`. The schema defined in the gateway configuration can differ from that in the upstream API, which enables fields to be restricted by default.
```
Comment (davegarvey) [inline tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: Should be a bullet point

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+Field-based permissions can also be used, to authorise client access of individual fields available in the schema. In this case, `id` and `name`.
```
Comment (davegarvey) [inline tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: Should be a bullet point

File: tyk-docs/content/api-management/manage-apis/api-security/overview.md (hunk 1 above)
Comment (davegarvey) [reply tyk-docs/content/api-management/manage-apis/api-security/overview.md]: Perhaps "API security best practices using API management"?

File: tyk-docs/content/api-management/manage-apis/api-security/resource-consumption.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+**Protect Failing Services**: Defend struggling endpoints by using a [circuit breaker]({{< ref "planning-for-production/ensure-high-availability/circuit-breakers#overview" >}}). This feature protects endpoints by detecting error responses, then blocking requests for a short duration to allow them to recover. The same principle can be applied in a wider sense by using [uptime tests]({{< ref "tyk-apis/tyk-gateway-api/api-definition-objects/uptime-tests" >}}), though this works on a host level instead, by removing failed hosts from the gateway load balancer.
+
+**Enforce Network-Level Security**: Problematic clients can be prevented from accessing the API by [blocking their address]({{< ref "tyk-apis/tyk-gateway-api/api-definition-objects/ip-blacklisting" >}}). Conversely, for APIs with a known set of clients, [allow lists]({{< ref "tyk-apis/tyk-gateway-api/api-definition-objects/ip-whitelisting" >}}) can be used to create a list of allow addresses, thereby implicitly blocking every other address from the API.
```
Comment (davegarvey) [inline tyk-docs/content/api-management/manage-apis/api-security/resource-consumption.md]: Perhaps "allow addresses" should be "allowed addresses". In the original doc it was "allow addresses", as per here, but having read it back I think it reads better as "allowed addresses". I've updated the original doc.

**Review by davegarvey — CHANGES_REQUESTED** [review]
I've made a few minor points. Otherwise, I think it's good to go.

File: tyk-docs/content/api-management/manage-apis/api-security/resource-consumption.md (hunk 5 above)
Comment (dcs3spp) [reply tyk-docs/content/api-management/manage-apis/api-security/resource-consumption.md]: done

File: tyk-docs/content/api-management/manage-apis/api-security/overview.md (hunk 1 above)
Comment (dcs3spp) [reply tyk-docs/content/api-management/manage-apis/api-security/overview.md]: done

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md (hunk 2 above)
Comment (dcs3spp) [reply tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: done

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md (hunk 3 above)
Comment (dcs3spp) [reply tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: Done

File: tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md (hunk 4 above)
Comment (dcs3spp) [reply tyk-docs/content/api-management/manage-apis/api-security/authorisation-levels.md]: Done

**Comment by letzya** [comment]
looks good! @davegarvey @dcs3spp should we also add a link to your OWASP blog?

File: tyk-docs/data/menu.yaml [hunk 6]
```diff
@@ -3688,17 +3688,46 @@ menu:
         category: Page
         show: True
   - title: "APIM Best Practice"
-    path: 
+    path: /api-management/manage-apis/api-security/overview
```
Comment (letzya) [inline tyk-docs/data/menu.yaml]: Are you sure this file path is correct?

File: tyk-docs/data/menu.yaml [hunk 7]
```diff
@@ -3688,17 +3688,46 @@ menu:
...
-      show: False
+      category: Directory
+      show: True
+      menu:
+      - title: "API Security Overview"
+        path: /api-management/manage-apis/api-security/overview
+        category: Page
+        show: True
+      - title: "Authentication"
+        path: /api-management/manage-apis/api-security/authentication
```
Comment (letzya) [inline tyk-docs/data/menu.yaml]: Shouldn't all the pages be under a certain directory?

**Review by letzya — CHANGES_REQUESTED** [review]
per comments

File: tyk-docs/data/menu.yaml (hunk 6 above)
Comment (dcs3spp) [reply tyk-docs/data/menu.yaml]: 🤔 Currently there is no content in the APIM section. Can add an small overview page explaining that this section relates to suggesting best practices for API management. And add link to security overview page in there?

File: tyk-docs/data/menu.yaml (hunk 7 above)
Comment (dcs3spp) [reply tyk-docs/data/menu.yaml]: if do that then that implies there will be other api-security content in future. If that is the case then could have this structure:

_/api-management/manage-apis/api-security/api-security-best-practices-using-apim/_ 

and have all the content files for the article in that folder?

File: tyk-docs/data/menu.yaml [hunk 8]
```diff
@@ -3688,17 +3688,46 @@ menu:
...
+    path: /apim-best-practice/overview
     category: Tab
-    show: False
+    show: True
     menu:
     - title: "Overview"
+      path: /apim-best-practice/overview
+      category: Page
+      show: True
+    - title: "API Security Best Practice"
```
Comment (letzya) [inline tyk-docs/data/menu.yaml:3699]:
```suggestion
    - title: "API Securitye"
```
```suggestion
    - title: "API Security"
```

File: tyk-docs/data/menu.yaml (hunk 6 above)
Comment (dcs3spp) [reply tyk-docs/data/menu.yaml]: Done

File: tyk-docs/data/menu.yaml (hunk 7 above)
Comment (dcs3spp) [reply tyk-docs/data/menu.yaml]: Done

**Pull request merged**